### PR TITLE
Use bounded queue for pact service

### PR DESCRIPTION
**Pull request by larskuhtz**
* [x] Replace `TQueue` with `TBQueue`.
* [ ] Add telemetry to monitor statistics of the size of the pact queue

### CODE REVIEW

File: src/Chainweb/Pact/Service/PactInProcApi.hs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
```
Comment (fosskers) [inline src/Chainweb/Pact/Service/PactInProcApi.hs]: Necessary?

File: test/Chainweb/Test/Pact/ChainData.hs [hunk 2]
```diff
@@ -157,14 +156,14 @@ withPact
...
-    -> (IO (TQueue RequestMsg) -> TestTree)
+    -> (IO PactQueue -> TestTree)
     -> TestTree
 withPact iopdb iobhdb mempool iodir f =
     withResource startPact stopPact $ f . fmap snd
   where
     startPact = do
         mv <- newEmptyMVar
-        reqQ <- atomically newTQueue
+        reqQ <- atomically (newTBQueue 2000)
```
Comment (fosskers) [inline test/Chainweb/Test/Pact/ChainData.hs]: Shouldn't this use the `pactQueueSize` value defined above?

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 3]
```diff
@@ -80,13 +79,13 @@ tests = ScheduledTest label $ withRocksResource $ \rocksIO ->
...
     -> MemPoolAccess
-    -> (IO (TQueue RequestMsg) -> TestTree)
+    -> (IO PactQueue -> TestTree)
     -> TestTree
 withPact rocksIO mempool f = withResource startPact stopPact $ f . fmap snd
   where
     startPact = do
         mv <- newEmptyMVar
-        reqQ <- atomically newTQueue
+        reqQ <- atomically (newTBQueue 2000)
```
Comment (fosskers) [inline test/Chainweb/Test/Pact/PactInProcApi.hs]: Here too.

File: test/Chainweb/Test/Pact/PactReplay.hs [hunk 4]
```diff
@@ -286,14 +285,14 @@ withPact
...
-    -> (IO (TQueue RequestMsg) -> TestTree)
+    -> (IO PactQueue -> TestTree)
     -> TestTree
 withPact logLevel iopdb iobhdb mempool iodir f =
     withResource startPact stopPact $ f . fmap snd
   where
     startPact = do
         mv <- newEmptyMVar
-        reqQ <- atomically newTQueue
+        reqQ <- atomically (newTBQueue 2000)
```
Comment (fosskers) [inline test/Chainweb/Test/Pact/PactReplay.hs]: Here too.

File: test/Chainweb/Test/Pact/TTL.hs [hunk 5]
```diff
@@ -117,14 +116,14 @@ withPact
...
-    -> (IO (TQueue RequestMsg) -> TestTree)
+    -> (IO PactQueue -> TestTree)
     -> TestTree
 withPact iopdb iobhdb mempool iodir f =
     withResource startPact stopPact $ f . fmap snd
   where
     startPact = do
         mv <- newEmptyMVar
-        reqQ <- atomically newTQueue
+        reqQ <- atomically (newTBQueue 2000)
```
Comment (fosskers) [inline test/Chainweb/Test/Pact/TTL.hs]: Here too.

**Comment by fosskers** [comment]
Otherwise looks good. If we attempt a `addRequest :: PactQueue -> RequestMsg -> IO ()` when the Queue is full, will it just block on the IO action?

File: src/Chainweb/Pact/PactService.hs [hunk 6]
```diff
@@ -284,20 +288,67 @@ serviceRequests memPoolAccess reqQ = do
...
+    tryOne which mvar m = tryOne' which mvar Right m
+
+    tryOne'
+        :: String
+        -> MVar (Either PactException b)
+        -> (a -> Either PactException b)
+        -> PactServiceM cas a
+        -> PactServiceM cas ()
+    tryOne' which mvar post m =
+        (evalPact (post <$> m) >>= (liftIO . putMVar mvar))
```
Comment (emilypi) [inline src/Chainweb/Pact/PactService.hs]: `evalPact` the name clashes with some Pact constructs that we use. Can we use a different name for this (e.g. `eval_`)?

File: src/Chainweb/Pact/Service/PactInProcApi.hs [hunk 7]
```diff
@@ -1,10 +1,12 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
```
Comment (emilypi) [inline src/Chainweb/Pact/Service/PactInProcApi.hs]: unused

File: test/Chainweb/Test/Pact/Utils.hs [hunk 8]
```diff
@@ -547,24 +547,22 @@ withPact
...
   where
     startPact = do
         mv <- newEmptyMVar
-        reqQ <- atomically newTQueue
+        reqQ <- atomically $ newTBQueue pactQueueSize
         pdb <- iopdb
         bhdb <- iobhdb
         dir <- iodir
-        a <- withLink $ initPactService version cid logger reqQ mempool mv
+        a <- async $ initPactService version cid logger reqQ mempool mv
```
Comment (emilypi) [inline test/Chainweb/Test/Pact/Utils.hs]: This isn't the only place `withLink` was used - we can remove it from `Chainweb.Utils`, and there are 2 more tests using it (`PactReplayTest` and `PactInProcApiTest` iirc)

File: src/Chainweb/Pact/Service/PactInProcApi.hs [hunk 9]
```diff
@@ -89,22 +90,29 @@ withPactService'
...
-        reqQ <- atomically (newTQueue :: STM (TQueue RequestMsg))
-        a <- withLink $
-             PS.initPactService ver cid logger reqQ memPoolAccess cdbv bhDb
-                                pdb dbDir nodeid resetDb
-        evaluate =<< rst (action reqQ) `finally` closeQueue reqQ `finally` wait a
+withPactService' ver cid logger memPoolAccess cdbv bhDb pdb dbDir nodeid resetDb action = do
+    reqQ <- atomically $ newTBQueue pactQueueSize
+    race (server reqQ) (client reqQ) >>= \case
+        Left () -> error "pact service terminated unexpectedly"
+        Right a -> return a
```
Comment (fosskers) [inline src/Chainweb/Pact/Service/PactInProcApi.hs]: Do we ever expect this to exit?

File: src/Chainweb/Pact/Service/PactInProcApi.hs [hunk 10]
```diff
@@ -89,22 +90,29 @@ withPactService'
...
+  where
+    client reqQ = action reqQ
+    server reqQ = runForever logg "pact-service" $
```
Comment (fosskers) [inline src/Chainweb/Pact/Service/PactInProcApi.hs]: Is `withPactService'` expected to run forever too?

File: src/Chainweb/Pact/Service/PactInProcApi.hs (hunk 9 above)
Comment (larskuhtz) [reply src/Chainweb/Pact/Service/PactInProcApi.hs]: Only the client (right side). I think we have test cases that just terminate and expect the server to shut down.

File: src/Chainweb/Pact/Service/PactInProcApi.hs (hunk 10 above)
Comment (larskuhtz) [reply src/Chainweb/Pact/Service/PactInProcApi.hs]: No, only the server component is expect to stay alive as long as the client is running.

In practice, in a production chainweb node, the client above will also run forever, but that's not a concern of this function.

**Pull request merged**